core/src/main/java/org/apache/iceberg/PartitionsTable.java

```diff
@@ -20,90 +20,140 @@
  */
 public class PartitionsTable extends BaseMetadataTable {
 
-  private final TableOperations ops;
-  private final Table table;
   private final Schema schema;
+  static final boolean PLAN_SCANS_WITH_WORKER_POOL =
+      SystemProperties.getBoolean(SystemProperties.SCAN_THREAD_POOL_ENABLED, true);
+
+  PartitionsTable(TableOperations ops, Table table) {
+    this(ops, table, table.name() + ".partitions");
+  }
+
+  PartitionsTable(TableOperations ops, Table table, String name) {
+    super(ops, table, name);
 
-  public PartitionsTable(TableOperations ops, Table table) {
-    this.ops = ops;
-    this.table = table;
     this.schema = new Schema(
         Types.NestedField.required(1, "partition", table.spec().partitionType()),
         Types.NestedField.required(2, "record_count", Types.LongType.get()),
         Types.NestedField.required(3, "file_count", Types.IntegerType.get())
     );
   }
 
-  @Override
-  Table table() {
-    return table;
-  }
-
-  @Override
-  String metadataTableName() {
-    return "partitions";
-  }
-
   @Override
   public TableScan newScan() {
-    return new PartitionsScan();
+    return new PartitionsScan(operations(), table());
   }
 
   @Override
   public Schema schema() {
+    if (table().spec().fields().size() < 1) {
+      return schema.select("record_count", "file_count");
+    }
     return schema;
   }
 
-  private DataTask task(TableScan scan) {
-    return StaticDataTask.of(
-        ops.io().newInputFile(ops.current().metadataFileLocation()),
-        partitions(table, scan.snapshot().snapshotId()),
-        PartitionsTable::convertPartition);
+  @Override
+  MetadataTableType metadataTableType() {
+    return MetadataTableType.PARTITIONS;
+  }
+
+  private DataTask task(StaticTableScan scan) {
+    TableOperations ops = operations();
+    Iterable<Partition> partitions = partitions(scan);
+    if (table().spec().fields().size() < 1) {
+      // the table is unpartitioned, partitions contains only the root partition
+      return StaticDataTask.of(
+          io().newInputFile(ops.current().metadataFileLocation()),
+          schema(), scan.schema(), partitions,
+          root -> StaticDataTask.Row.of(root.recordCount, root.fileCount)
+      );
+    } else {
+      return StaticDataTask.of(
+          io().newInputFile(ops.current().metadataFileLocation()),
+          schema(), scan.schema(), partitions,
+          PartitionsTable::convertPartition
+      );
+    }
   }
 
   private static StaticDataTask.Row convertPartition(Partition partition) {
     return StaticDataTask.Row.of(partition.key, partition.recordCount, partition.fileCount);
   }
 
-  private static Iterable<Partition> partitions(Table table, Long snapshotId) {
-    PartitionSet partitions = new PartitionSet();
-    TableScan scan = table.newScan();
+  private static Iterable<Partition> partitions(StaticTableScan scan) {
+    CloseableIterable<FileScanTask> tasks = planFiles(scan);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the following resource: *tasks*. Execution paths that do not contain closure statements were detected. Ensure *tasks* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).